File: NestedConditionals/Python/Correct/reservation.py
import datetime

from customer import Customer


class Reservation:
    def __init__(self, customer: Customer, date_time: datetime) -> None:
        self.customer = customer
        self.date_time = date_time
        self.is_canceled: bool = False

    def cancel(self) -> bool:
        MESSAGE_ERROR = "It's too late to cancel."
        
        if (self.is_cancellation_period_over()):
            raise Exception(MESSAGE_ERROR)
            
        self.is_canceled = True
    
    def is_cancellation_period_over(self):
        return (self.customer.IsGoldCustomer() and self.less_than(24) or
                not self.customer.IsGoldCustomer() and self.less_than(48))
    
    def less_than(self, maxHours):
        return ((self.date_time - datetime.now()).TotalHours < maxHours)
    

File: NestedConditionals/C#/Correct/Reservation.cs
namespace NestedConditionals
{
    public class Reservation
    {
        public DateTime From { get; set; }
        public Customer Customer { get; set; }
        public bool IsCanceled { get; set; }

        public Reservation(Customer customer, DateTime dateTime)
        {
            From = dateTime;
            Customer = customer;
        }

        public void Cancel()
        {
            if (IsCancellationPeriodOver())
                throw new InvalidOperationException("It's too late to cancel.");

            IsCanceled = true;
        }

        private bool IsCancellationPeriodOver() =>
            Customer.IsGoldCustomer() && LessThan(24) || !Customer.IsGoldCustomer() && LessThan(48);

        private bool LessThan(int maxHours) =>
            ((From - DateTime.Now).TotalHours < maxHours);
    }
}